File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/submodel/metamodel/connected/submodelelement/property/ConnectedProperty.java
package org.eclipse.basyx.submodel.metamodel.connected.submodelelement.property;

import java.util.Map;

import org.eclipse.basyx.submodel.metamodel.api.submodelelement.property.IProperty;
import org.eclipse.basyx.submodel.metamodel.api.submodelelement.property.PropertyType;
import org.eclipse.basyx.submodel.metamodel.connected.submodelelement.ConnectedDataElement;
import org.eclipse.basyx.submodel.metamodel.map.submodelelement.property.SingleProperty;
import org.eclipse.basyx.vab.modelprovider.VABElementProxy;
/**
 * "Connected" implementation of IProperty
 * @author rajashek
 *
 */
public abstract class ConnectedProperty extends ConnectedDataElement implements IProperty {
	private PropertyType type;

	public ConnectedProperty(PropertyType type, VABElementProxy proxy) {
		super(proxy);		
		this.type = type;
	}

	@Override
	public PropertyType getPropertyType() {
		return type;
	}

	@Override
	public void setValueId(String obj) {
		getProxy().setModelPropertyValue(SingleProperty.VALUEID, obj);
		
	}

	@Override
	public String getValueId() {
		return (String) getProxy().getModelPropertyValue(SingleProperty.VALUEID);
	}

	@SuppressWarnings("unchecked")
	protected <T> T retrieveObject() {
		return (T) ((Map<String, Object>) getProxy().getModelPropertyValue(SingleProperty.VALUE))
				.get(SingleProperty.VALUE);
	}
}

File: sdks/c++/basys.sdk.cc/src/submodel/submodel/connected/submodelelement/property/ConnectedProperty.cpp
/*
 * ConnectedProperty.cpp
 *
 *      Author: wendel
 */
#include "ConnectedProperty.h"
#include "basyx/anyTypeChecker.h"

namespace basyx {
namespace submodel {


ConnectedProperty::ConnectedProperty(PropertyType type, std::shared_ptr<vab::core::proxy::IVABElementProxy> proxy) :
  ConnectedDataElement(proxy),
  type(type)
{}

IProperty::PropertyType ConnectedProperty::getPropertyType() const
{
  return this->type;
}

void ConnectedProperty::setId(const std::string & id)
{
  this->setId(id);
}

std::string ConnectedProperty::getId() const
{
  return this->getId();
}

basyx::object ConnectedProperty::retrieveObject() const
{
  return this->getProxy()->readElementValue(IProperty::Path::Value).Get<basyx::object::object_map_t>().at(IProperty::Path::Value);
}

void ConnectedProperty::setValueId(const std::string & valueId)
{}

std::string ConnectedProperty::getValueId() const
{
  return std::string();
}

}
}